experimental support for query batching (#3837)
apollographql/router: docs/source/executing-operations/query-batching.mdx

---
title: Query batching
---

The Apollo Router provides experimental support for Query Batching.

## Configuration

By default this feature is disabled. If enabled, a single mode of operation is supported, `batch_http_link`.

```yaml title="router.yaml"
experimental_batching:
  enabled: true
  mode: batch_http_link
```

If a router receives a query batch from a client and batching is *not* enabled, the client will receive an error which clearly identifies this as `BATCHING_NOT_ENABLED`.

## Metrics

<table class="field-table metrics">
  <thead>
    <tr>
      <th>Name</th>
      <th>Attributes</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td style="min-width: 150px;">

##### `apollo.router.operations.batching`

</td>
<td>

mode

</td>
<td>

counter for the number of batched requests received

</td>
</tr>

<tr>
<td style="min-width: 150px;">

> ##### `apollo.router.config.batching`

Author: done: 3ba599f6c5ed0d9d62b01d45cb20e1695f239fc9

```graphql
[
query MyFirstQuery {
  me {
    id
  }
},,
query MySecondQuery {
  me {
    name
  }
}
]
```

As a result, an error is returned for the invalid batch:

```json
{"errors":
  [
    {"message":"Invalid GraphQL request","extensions":{"details":"failed to deserialize the request body into JSON: expected value at line 1 column 54","code":"INVALID_GRAPHQL_REQUEST"}}
  ]
}
```

### Individual query error

If a single query in a batch cannot be processed, this will result in an individual error. 

For example, the query `MyFirstQuery` is accessing a field that doesn't exist, while the rest of the batch query is valid.

```graphql
[
query MyFirstQuery {
  me {
    thisfielddoesnotexist
  }
},
query MySecondQuery {
  me {
    name
  }
}
]
```

As a result, an error is returned for the individual query:

```json
[
  {"errors":
    [
      {"message":"cannot query field 'thisfielddoesnotexist' on type 'User'",
       "extensions":{"type":"User","field":"thisfielddoesnotexist","code":"INVALID_FIELD"}
      }
    ]
  },
  {"data":{"me":{"name":"Ada Lovelace"}}}
]
```

## Known limitations

### Unsupported query modes
 

> When batching is enabled, any mode of querying that results in a stream of responses is unsupported, including:

Reviewer: You mentioned `@delay` in the changelog

Author: I updated this to `@defer` in [33f5934](https://github.com/apollographql/router/pull/3837/commits/33f593473b6bd3989d3bdb5151c9a7f16a98c4e4)

---
title: Query batching
description: Receive query batches with the Apollo Router  
---

> **Experimental Feature**  
> The Apollo Router provides *experimental* support for client query batching.

Learn about query batching and how to configure the Apollo Router to receive query batches.

## About query batching

Modern applications use several operations to render a single page. This is usually the result of a component-based architecture where individual micro-frontends (MFE) make requests separately to fetch data relevant to them. Not only does this cause a performance overhead (i.e. different components may be requesting the same data), it can also cause a consistency issue. To combat this, MFE-based UIs batch multiple client queries and operations, issued close together, into one HTTP request. This is supported in Apollo Client and Apollo Server.

> The Apollo Router now supports client request batching. If you’re using Apollo Client, you can leverage the built-in support for batching to reduce the number of individual requests sent to the router.

Author: No we don't. I've removed "now"

The Apollo Router must be configured to receive batch requests, otherwise it rejects them. When processing a batch request, the router deserializes and processes each operation of a batch independently, and it responds to the client only after all operations of the batch have been completed. Each operation executes concurrently with respect to other operations in the batch.

## Configure query batching

Both a client and the Apollo Router need to be configured to support query batching.

### Configure router

By default, receiving client request batches is disabled in Apollo Router. 

To enable query batching, set the following fields in your `router.yaml` configuration file:

```yaml title="router.yaml"
experimental_batching:
  enabled: true
  mode: batch_http_link
```

| Attribute | Description | Valid Values | Default Value |
| :-- | :-- | :-- | :-- |
| `enabled` | Flag to enable reception of client batches | boolean | `false` |
| `mode` | Supported client batching mode | `batch_http_link`:  the client uses Apollo Link and its [`BatchHttpLink`](/react/api/link/apollo-link-batch-http) link. | No Default |

### Configure client

> To enable batching in a client, configure `BatchHttpLink`. For details on implementing `BatchHttpLink`, see [batching operations](https://www.apollographql.com/docs/react/api/link/apollo-link-batch-http/).

Author: Well, that's a specific Apollo React client feature. It's not a requirement to make use of batching, but we call it out here because we know that we have customers/users using this client and who want to make use of batching. Perhaps:

To enable batching in an Apollo React client, configure `BatchHttpLink`. For details on implementing `BatchHttpLink`, see [batching operations](/react/api/link/apollo-link-batch-http/).

?

### Configuration compatibility

> If the router receives a query batch from a client, and batching is *not* enabled, the router sends a `BATCHING_NOT_ENABLED` error to the client, which clearly identifies why the query batch failed.

Author: The error does contain additional information. It would look something like this:
```
{
  "errors": [
    {
      "message": "Invalid GraphQL request",
      "extensions": {
        "details": "batching not enabled",
        "code": "BATCHING_NOT_ENABLED"
      }
    }
  ]
}
```
However, I agree that the dependent clause is superfluous.

## Metrics for query batching

> Metrics in the Apollo Router for query batching:

Author: It depends how your router is configured. For example, if you enable prometheus then you can see the metrics in prometheus. If you configure DataDog, then the metrics are exported to DataDog. You may also configure any other metrics ingrestion mechanism that understand OpenTelemetry.

I think that anyone who is interested in consuming metrics from the router will know this. We could add links to:
https://www.apollographql.com/docs/router/configuration/metrics/#using-prometheus but that's not telling the full story.

> Metrics in the Apollo Router for query batching:

Reviewer 2: Got it. We can leave as is. :)

<table class="field-table metrics">
  <thead>
    <tr>
      <th>Name</th>
      <th>Attributes</th>
      <th>Description</th>
    </tr>
  </thead>

<tbody>
<tr class="required">
<td style="min-width: 150px;">

##### `apollo.router.operations.batching`

</td>
<td>

mode

</td>
<td>

> Counter for the number received batch requests.

Reviewer 2: ```suggestion
Counter for the number of received batch requests.
```

</td>
</tr>

<tr class="required">
<td style="min-width: 150px;">

##### `apollo.router.operations.batching.size`

</td>
<td>

mode

</td>
<td>

Histogram for the size of received batch requests.

</td>
</tr>
</tbody>
</table>

## Query batch formats

### Request format

> 

Reviewer 2: ```suggestion

A batched request is an array of operations.

```

```graphql
[
query MyFirstQuery {
  me {
    id
  }
},
query MySecondQuery {
  me {
    name
  }
}
]
```

### Response format

> Responses are provided in JSON array, with ordering of responses matching the supplied request.

Reviewer 2: ```suggestion
Responses are provided in JSON array, with the order of responses matching the order of operations in the batched request.
```

```json
[
  {"data":{"me":{"id":"1"}}},
  {"data":{"me":{"name":"Ada Lovelace"}}}
]
```

## Error handling for query batching

### Batch error

> If a batch of queries cannot be processed, the entire batch will fail. 

Reviewer 2: ```suggestion
If a batch of queries cannot be processed, the entire batch fails.
```

> For example, this query is invalid because it has two commas to separate the batch of queries:

Reviewer 2: ```suggestion
For example, this batch request is invalid because it has two commas to separate the constituent queries:
```

```graphql
[
query MyFirstQuery {
  me {
    id
  }
},,
query MySecondQuery {
  me {
    name
  }
}
]
```

> As a result, an error is returned for the invalid batch:

Reviewer 2: ```suggestion
As a result, the router returns an invalid batch error:
```

```json
{"errors":
  [
    {"message":"Invalid GraphQL request","extensions":{"details":"failed to deserialize the request body into JSON: expected value at line 1 column 54","code":"INVALID_GRAPHQL_REQUEST"}}
  ]
}
```

### Individual query error

> If a single query in a batch cannot be processed, this will result in an individual error. 

Reviewer 2: ```suggestion
If a single query in a batch cannot be processed, this results in an individual error. 
```

For example, the query `MyFirstQuery` is accessing a field that doesn't exist, while the rest of the batch query is valid.

```graphql
[
query MyFirstQuery {
  me {
    thisfielddoesnotexist
  }
},
query MySecondQuery {
  me {
    name
  }
}
]
```

> As a result, an error is returned for the individual query:

Reviewer 2: ```suggestion
As a result, an error is returned for the individual invalid query and the other (valid) query returns a response.
```

```json
[
  {"errors":
    [
      {"message":"cannot query field 'thisfielddoesnotexist' on type 'User'",
       "extensions":{"type":"User","field":"thisfielddoesnotexist","code":"INVALID_FIELD"}
      }
    ]
  },
  {"data":{"me":{"name":"Ada Lovelace"}}}
]
```

## Known limitations

### Unsupported query modes
 

> When batching is enabled, any mode of querying that results in a stream of responses is unsupported, including:

Reviewer 2: Does this mean that the router cannot process these types of operations _at all_ or just in batches? I assume the latter, but let's make it clear, either way.

Author: You assume correctly. How about this?:
```
When batching is enabled, any batch operation that results in a stream of responses is unsupported, including:
```
I've made that edit. If you don't like it, feel free to suggest improvements.

Reviewer 2: That looks good, thank you! 🙇‍♀️

> - `@defer`

Reviewer 2: ```suggestion
- [`@defer`](/graphos/operations/defer/)
```

> - subscriptions

Reviewer 2: ```suggestion
- [subscriptions](/graphos/operations/subscriptions/)
```

## Configure query batching

Both the Apollo Router and client need to be configured to support query batching.

### Configure router

By default, receiving client query batches is _not_ enabled in the Apollo Router. 

To enable query batching, set the following fields in your `router.yaml` configuration file:

```yaml title="router.yaml"
experimental_batching:
  enabled: true
  mode: batch_http_link
```

| Attribute | Description | Valid Values | Default Value |
| :-- | :-- | :-- | :-- |
| `enabled` | Flag to enable reception of client query batches | boolean | `false` |
| `mode` | Supported client batching mode | `batch_http_link`:  the client uses Apollo Link and its [`BatchHttpLink`](/react/api/link/apollo-link-batch-http) link. | No Default |

### Configure client

To enable batching in an Apollo client, configure `BatchHttpLink`. For details on implementing `BatchHttpLink`, see [batching operations](/react/api/link/apollo-link-batch-http/).

### Configuration compatibility

If the router receives a query batch from a client, and batching is *not* enabled, the router sends a `BATCHING_NOT_ENABLED` error to the client.

> ## Metrics for query batching

Reviewer: I think we should at least mention them in the metrics docs where we list all metrics
